go/k8s/controllers/user_controller.go

```diff
@@ -100,66 +142,144 @@ func (r *UserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
 	defer cancel()
 
 	var (
-		err           error
-		backendURL    string
-		userID        int64
-		userState     UserState
-		userStatus    backendv1.UserStatus
-		retrievedUser *backendv1.User
+		err              error
+		backendURL       string
+		userID           int64
+		trustedDeveloper bool
+		userState        UserState
+		userStatus       backendv1.UserStatus
+		retrievedUser    *backendv1.User
 	)
 
 	defer func() {
 		if err != nil {
 			userStatus.State = err.Error()
 		}
 		if retrievedUser != nil {
+			if r.almondConfig.EnableDeveloperBackend && trustedDeveloper {
+				retrievedUser.Spec.Mode = "developer"
+			} else {
+				retrievedUser.Spec.Mode = "shared"
+			}
+			r.Update(ctx, retrievedUser)
 			retrievedUser.Status = userStatus
 			r.Status().Update(ctx, retrievedUser)
 			r.Log.Info("update status:", "user", retrievedUser.Spec.ID, "status", retrievedUser.Status)
 		}
 		r.Log.Info("--- end ---")
 	}()
 
-	backendURL, userID, err = r.getBackendURL(ctx, req.NamespacedName.Name, req)
+	userID, trustedDeveloper, err = r.getUserFromName(req.Name)
 	if err != nil {
-		return ctrl.Result{RequeueAfter: 2 * time.Second}, err
+		return ctrl.Result{}, err
 	}
+
+	if r.almondConfig.EnableDeveloperBackend && trustedDeveloper {
+		// for developer users, make sure deployment and service are up before proceeding.
+		user := &backendv1.User{}
+		if err = r.Client.Get(ctx, req.NamespacedName, user); err != nil {
+			if apierrors.IsNotFound(err) {
+				if err = r.deleteDeploymentService(ctx, req, userID); err != nil {
+					r.Log.Error(err, "fail to delete developer deployment or service")
+				}
+			}
+			return ctrl.Result{}, err
+		}
+		retrievedUser = user
+		if !retrievedUser.ObjectMeta.DeletionTimestamp.IsZero() {
+			// user is marked for deletion
+			user.Status.State = string(Stopping)
+			if err = r.deleteDeploymentService(ctx, req, userID); err != nil {
```

Comment:
Do we need to delete the deployment and service when we delete the user? What if we scale it to 0 instead?